File: app/src/main/java/de/westnordost/streetcomplete/about/AboutFragment.java
package de.westnordost.streetcomplete.about;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.preference.PreferenceFragmentCompat;

import de.westnordost.streetcomplete.ApplicationConstants;
import de.westnordost.streetcomplete.BuildConfig;
import de.westnordost.streetcomplete.FragmentContainerActivity;
import de.westnordost.streetcomplete.R;

public class AboutFragment extends PreferenceFragmentCompat
{


	@Override public void onCreatePreferences(Bundle savedInstanceState, String rootKey)
	{
		addPreferencesFromResource(R.xml.about);

		findPreference("version").setSummary(BuildConfig.VERSION_NAME);

		findPreference("license").setOnPreferenceClickListener( preference ->
		{
			Intent browserIntent = new Intent(Intent.ACTION_VIEW,
					Uri.parse("https://www.gnu.org/licenses/gpl-3.0.html"));
			startActivity(browserIntent);
			return true;
		});

		findPreference("authors").setOnPreferenceClickListener(preference ->
		{
			getFragmentActivity().setCurrentFragment(new CreditsFragment());
			return true;
		});

		findPreference("privacy").setOnPreferenceClickListener(preference ->
		{
			Fragment f = ShowHtmlFragment.create(
					getResources().getString(R.string.privacy_html) +
					getString(R.string.privacy_html_tileserver) +
					getString(R.string.privacy_html_third_party_quest_sources) +
					getString(R.string.privacy_html_image_upload),
					R.string.about_title_privacy_statement);
			getFragmentActivity().setCurrentFragment(f);
			return true;
		});

		findPreference("repository").setOnPreferenceClickListener(preference ->
		{
			Intent browserIntent = new Intent(Intent.ACTION_VIEW,
					Uri.parse("https://github.com/westnordost/StreetComplete/"));
			startActivity(browserIntent);
			return true;
		});

		findPreference("report_error").setOnPreferenceClickListener(preference ->
		{
			Intent browserIntent = new Intent(Intent.ACTION_VIEW,
					Uri.parse("https://github.com/westnordost/StreetComplete/issues/"));
			startActivity(browserIntent);
			return true;
		});

		findPreference("email_feedback").setOnPreferenceClickListener(preference ->
		{
			Intent intent = new Intent(Intent.ACTION_SENDTO);
			intent.setData(Uri.parse("mailto:"));
			intent.putExtra(Intent.EXTRA_EMAIL, new String[]{"osm@westnordost.de"});
			intent.putExtra(Intent.EXTRA_SUBJECT, ApplicationConstants.USER_AGENT + " Feedback");
			if (intent.resolveActivity(getActivity().getPackageManager()) != null)
			{
				startActivity(intent);
				return true;
			}
			return false;
		});
	}

	@Override public void onStart()
	{
		super.onStart();
		getActivity().setTitle(R.string.action_about);
	}

	private FragmentContainerActivity getFragmentActivity()
	{
		return (FragmentContainerActivity) getActivity();
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/about/AboutFragment.kt
package de.westnordost.streetcomplete.about

import android.content.ActivityNotFoundException
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.DrawableRes
import androidx.appcompat.app.AlertDialog
import androidx.core.net.toUri
import androidx.core.widget.TextViewCompat
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.recyclerview.widget.RecyclerView
import de.westnordost.streetcomplete.ApplicationConstants
import de.westnordost.streetcomplete.BuildConfig
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.view.ListAdapter
import kotlinx.android.synthetic.main.cell_labeled_icon_select_right.view.*
import java.util.*

/** Shows the about screen */
class AboutFragment : PreferenceFragmentCompat() {

    interface Listener {
        fun onClickedChangelog()
        fun onClickedCredits()
        fun onClickedPrivacyStatement()
    }
    private val listener: Listener? get() = parentFragment as? Listener ?: activity as? Listener

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.about)

        findPreference<Preference>("version")?.summary = getString(R.string.about_summary_current_version, "v" + BuildConfig.VERSION_NAME)
        findPreference<Preference>("version")?.setOnPreferenceClickListener {
            listener?.onClickedChangelog()
            true
        }

        findPreference<Preference>("license")?.setOnPreferenceClickListener {
            openUrl("https://www.gnu.org/licenses/gpl-3.0.html")
        }

        findPreference<Preference>("authors")?.setOnPreferenceClickListener {
            listener?.onClickedCredits()
            true
        }

        findPreference<Preference>("privacy")?.setOnPreferenceClickListener {
            listener?.onClickedPrivacyStatement()
            true
        }

        findPreference<Preference>("repository")?.setOnPreferenceClickListener {
            openUrl("https://github.com/westnordost/StreetComplete/")
        }

        findPreference<Preference>("translate")?.summary = resources.getString(
            R.string.about_description_translate,
            Locale.getDefault().displayLanguage,
            resources.getInteger(R.integer.translation_completeness)
        )
        findPreference<Preference>("translate")?.setOnPreferenceClickListener {
            openUrl("https://poeditor.com/join/project/IE4GC127Ki")
        }

        findPreference<Preference>("report_error")?.setOnPreferenceClickListener {
            openUrl("https://github.com/westnordost/StreetComplete/issues/")
        }

        findPreference<Preference>("email_feedback")?.setOnPreferenceClickListener {
            sendFeedbackEmail()
        }

        findPreference<Preference>("rate")?.isVisible = isInstalledViaGooglePlay()
        findPreference<Preference>("rate")?.setOnPreferenceClickListener {
            openGooglePlayStorePage()
        }

        findPreference<Preference>("donate")?.setOnPreferenceClickListener {
            showDonateDialog()
            true
        }
    }

    override fun onStart() {
        super.onStart()
        activity?.setTitle(R.string.action_about2)
    }

    private fun isInstalledViaGooglePlay(): Boolean {
        val appCtx = context?.applicationContext ?: return false
        val installerPackageName = appCtx.packageManager.getInstallerPackageName(appCtx.packageName)
        return installerPackageName == "com.android.vending"
    }

    private fun openUrl(url: String): Boolean {
        val intent = Intent(Intent.ACTION_VIEW, url.toUri())
        return tryStartActivity(intent)
    }

    private fun sendFeedbackEmail(): Boolean {
        val intent = Intent(Intent.ACTION_SENDTO)
        intent.data = "mailto:".toUri()
        intent.putExtra(Intent.EXTRA_EMAIL, arrayOf("osm@westnordost.de"))
        intent.putExtra(Intent.EXTRA_SUBJECT, ApplicationConstants.USER_AGENT + " Feedback")
        return tryStartActivity(intent)
    }

    private fun openGooglePlayStorePage(): Boolean {
        val appPackageName = context?.applicationContext?.packageName ?: return false
        return openUrl("market://details?id=$appPackageName")
    }

    private fun tryStartActivity(intent: Intent): Boolean {
        return try {
            startActivity(intent)
            true
        } catch (e: ActivityNotFoundException) {
            false
        }
    }

    private fun showDonateDialog() {
        val ctx = context ?: return

        val view = LayoutInflater.from(ctx).inflate(R.layout.dialog_donate, null)
        val listView = view.findViewById<RecyclerView>(R.id.donateList)
        listView.adapter = DonationPlatformAdapter(DonationPlatform.values().asList())

        AlertDialog.Builder(ctx)
                .setTitle(R.string.about_title_donate)
                .setView(view)
                .show()
    }

    private inner class DonationPlatformAdapter(list: List<DonationPlatform>): ListAdapter<DonationPlatform>(list) {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
                ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.cell_labeled_icon_select_right, parent, false))

        inner class ViewHolder(itemView: View) : ListAdapter.ViewHolder<DonationPlatform>(itemView) {
            override fun onBind(with: DonationPlatform) {
                itemView.imageView.setImageResource(with.iconId)
                itemView.textView.text = with.title
                itemView.setOnClickListener { openUrl(with.url) }
                TextViewCompat.setTextAppearance(itemView.textView, R.style.TextAppearance_Title)
            }
        }
    }
}

private enum class DonationPlatform(val title: String, @DrawableRes val iconId: Int, val url: String) {
    GITHUB("GitHub Sponsors", R.drawable.ic_github, "https://github.com/sponsors/westnordost"),
    LIBERAPAY("Liberapay", R.drawable.ic_liberapay, "https://liberapay.com/westnordost"),
    PATREON("Patreon", R.drawable.ic_patreon, "https://patreon.com/westnordost")
}
